python/whylogs/core/constraints/metric_constraints.py

```diff
@@ -171,26 +174,297 @@ def validate_profile(self, dataset_profile: DatasetProfileView) -> Tuple[bool, O
+class DatasetConstraint:
+    """
+    Implements dataset-level constraints that are not attached to a specific
+    metric or column. The condition Callable takes the DatasetProfileView as
+    input and returns a boolean indicating whether the condition is satisfied
+    as well as a dictionary mapping 'column_name/metric_namespace' -> Metric for
+    any metrics the condition used during its evaluation.
+    """
+
+    condition: Callable[[DatasetProfileView], Tuple[bool, Dict[str, Metric]]]
+    name: str
+    require_column_existence: bool = True  # Applies to all columns referenced in the constraint
+
+    def _get_metric_summary(self, metrics: Dict[str, Metric]) -> Optional[Dict[str, Any]]:
+        summary: Dict[str, Any] = dict()
+        for path, metric in metrics.items():
+            metric_summary = pretty_display(metric.namespace, metric.to_summary_dict())
+            for component, value in metric_summary.items():
+                summary[f"{path}/{component}"] = value
+        return summary
+
+    def validate_profile(self, dataset_profile: DatasetProfileView) -> Tuple[bool, Optional[Dict[str, Any]]]:
+        try:
+            validate_result, metrics = self.condition(dataset_profile)
+        except MissingMetric as e:
+            if self.require_column_existence:
+                logger.info(f"validate_profile could not get metric {str(e)} so returning False.")
+                return (False, None)
+
+            logger.info(
+                f"validate_profile could not get metric {str(e)}  "
+                f"but require_column_existence is False, so returning True. "
+            )
+            return (True, None)
+
+        metric_summary = self._get_metric_summary(metrics)
+        return (validate_result, metric_summary)
+
+
+class MetricConstraintWrapper:
+    """
+    Converts a MetricConstraint not associated with a specific column
+    into a Callable that can be a DatasetConstraint condition.
+    """
+
+    def __init__(self, constraint: MetricConstraint) -> None:
+        self.constraint = constraint
+
+    def __call__(self, profile: DatasetProfileView) -> Tuple[bool, Dict[str, Metric]]:
+        valid, _ = self.constraint.validate_profile(profile)
+        metrics = self.constraint.metric_selector.apply(profile)  # Metrics used during evaluation
+        columns = profile.get_columns()
+        metric_map: Dict[str, Metric] = dict()
+        # Find the columns the referenced Metrics live in
+        for metric in metrics:
+            for column_name, column_view in columns.items():
+                if metric in column_view.get_metrics():
+                    metric_map[f"{column_name}/{metric.namespace}"] = metric
+                    break
+        return valid, metric_map
+
+
+_INT_RE = re.compile(r"[-+]?\d+")
+_FLOAT_RE = re.compile(r"[-+]?(\d+(\.\d*)?|\.\d+)([eE][-+]?\d+)?")
+
+
+class PrefixCondition:
+    """
+    Interpret expressions in the form used to serialize Predicate expressions.
```

Comment:
Maybe we could add the information that it uses prefix (polish) notation, and add the link:

https://en.wikipedia.org/wiki/Polish_notation

Reply:
I'm not sure it's good as the user interface... I'm thinking of it more as a serialization format. I tried a version using operator overloading, which sort of worked but had some infelicities. I think a fluent API would be awkward with the arithmetic operators. @labzhinova suggested an expression wizard UI in the front end. That would be nice, but doesn't answer how to write them in whylogs. Still pondering :)